### PR TITLE
Add custom material files, shared materials, automatic GC of materials

### PR DESCRIPTION
**Pull request by sin3point14**
emoji please?

### CODE REVIEW

File: rootex/core/renderer/material_library.cpp [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+{
+	return s_Materials;
+}
+void MaterialLibrary::SaveAll()
+{
+	for (auto& [materialName, materialInfo] : s_Materials)
+	{
+		if (Ref<Material> lockedMaterial = s_Materials["DEFAULT"].second.lock())
+		{
+			TextResourceFile* materialFile = ResourceLoader::CreateNewTextResourceFile("game/assets/materials/" + materialName);
```
Comment (IronicallySerious) [inline rootex/core/renderer/material_library.cpp:82]: Do not use `ResourceLoader::CreateNewTextResourceFile()`, use `ResourceLoader::CreateTextResourceFile()` instead. You are trying to gain access to a cached material file, but the CreateNewTextResourceFile() function creates a new text file (`> touch file_name.txt`)

File: rootex/core/renderer/material_library.cpp [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+			materialFile->putString(lockedMaterial->getJSON());
+		}
+	}
+}
+
+Ref<Material> MaterialLibrary::CreateNewMaterialFile(String& materialName, String& materialType)
+{
+	if (materialName == "DEFAULT")
```
Comment (IronicallySerious) [inline rootex/core/renderer/material_library.cpp:90]: Default

File: rootex/core/renderer/material_library.cpp [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+		return nullptr;
+	}
+	materialName += ".rmat";
+	if (s_Materials.find(materialName) == s_Materials.end())
+	{
+		TextResourceFile* materialFile = ResourceLoader::CreateNewTextResourceFile("game/assets/materials/" + materialName);
+		Ref<Material> material(s_MaterialDatabase[materialType].first());
+		materialFile->putString(material->getJSON());
+		s_Materials[materialName] = { materialType, material };
+		//m_MaterialList.push_back(materialName);
```
Comment (IronicallySerious) [inline rootex/core/renderer/material_library.cpp:101]: Remove unused code

File: rootex/core/renderer/material_library.h [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+	typedef Material* (*MaterialDefaultCreator)();
+	typedef Material* (*MaterialCreator)(const JSON::json& materialDescription);
+	typedef HashMap<String, std::pair<MaterialDefaultCreator, MaterialCreator>> MaterialDatabase;
+	static MaterialDatabase s_MaterialDatabase;
+public:
+	static void LoadMaterials();
+	static Ref<Material> GetMaterial(String& materialName);
+	static Ref<Material> GetDefaultMaterial();
+#ifdef ROOTEX_EDITOR
```
Comment (IronicallySerious) [inline rootex/core/renderer/material_library.h:23]: Newline before this

File: rootex/core/renderer/material_library.h [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+	static MaterialMap& GetAllMaterials();
+	void SaveAll();
+	static Ref<Material> CreateNewMaterialFile(String& materialName, String& materialType);
+	//static Vector<String> m_MaterialList;
```
Comment (IronicallySerious) [inline rootex/core/renderer/material_library.h:27]: Remove unused code

File: rootex/core/renderer/materials/color_material.cpp [hunk 6]
```diff
@@ -1,10 +1,11 @@
 #include "color_material.h"
 
-#include "renderer/shader_library.h"
 #include "framework/systems/render_system.h"
+#include "renderer/shader_library.h"
 
-ColorMaterial::ColorMaterial()
+ColorMaterial::ColorMaterial(Color color): m_Color(color)
```
Comment (IronicallySerious) [inline rootex/core/renderer/materials/color_material.cpp:6]: Tab

File: rootex/core/renderer/materials/color_material.cpp [hunk 7]
```diff
@@ -1,10 +1,11 @@
...
 {
+	m_TypeName = "Color Material";
```
Comment (IronicallySerious) [inline rootex/core/renderer/materials/color_material.cpp:8]: Better do this in the constructor initialization list

File: rootex/core/renderer/materials/color_material.cpp [hunk 8]
```diff
@@ -19,8 +20,43 @@ void ColorMaterial::setVSConstantBuffer(const VSSolidConstantBuffer& constantBuf
...
+	j["color"]["r"] = m_Color.x;
+	j["color"]["g"] = m_Color.y;
+	j["color"]["b"] = m_Color.z;
+	j["color"]["a"] = m_Color.w;
+	return j;
+}
+
+#ifdef ROOTEX_EDITOR
+#include "imgui.h"
+#include "imgui_stdlib.h"
```
Comment (IronicallySerious) [inline rootex/core/renderer/materials/color_material.cpp:56]: No need for this

File: rootex/core/renderer/materials/color_material.h [hunk 9]
```diff
@@ -16,11 +16,22 @@ class ColorMaterial : public Material
 		End
 	};
 
-	ColorMaterial();
+	ColorMaterial(Color color);
```
Comment (IronicallySerious) [inline rootex/core/renderer/materials/color_material.h:19]: Take in a `const Color&`

File: rootex/core/renderer/materials/textured_material.cpp [hunk 10]
```diff
@@ -1,18 +1,27 @@
...
 #include "renderer/shader_library.h"
 #include "framework/systems/light_system.h"
 #include "framework/systems/render_system.h"
 #include "renderer/texture.h"
 
-TexturedMaterial::TexturedMaterial(Ref<Texture> diffuseTexture)
+TexturedMaterial::TexturedMaterial(String& imagePath)
     : Material(ShaderLibrary::GetDiffuseShader())
-    , m_DiffuseTexture(diffuseTexture)
+    //, m_DiffuseTexture(diffuseTexture)
```
Comment (IronicallySerious) [inline rootex/core/renderer/materials/textured_material.cpp:12]: Unused code

File: rootex/core/renderer/materials/textured_material.cpp [hunk 11]
```diff
@@ -1,18 +1,27 @@
...
     , m_DiffuseShader(reinterpret_cast<DiffuseShader*>(m_Shader))
+    //, m_ImageFile(imageRes)
```
Comment (IronicallySerious) [inline rootex/core/renderer/materials/textured_material.cpp:14]: Unused

File: rootex/core/renderer/materials/textured_material.cpp [hunk 12]
```diff
@@ -1,18 +1,27 @@
...
 {
+	m_TypeName = "Textured Material";
+	m_ImageFile = ResourceLoader::CreateImageResourceFile(imagePath);
+	setTexture(m_ImageFile);
 	m_SamplerState = RenderingDevice::GetSingleton()->createSamplerState();
 	m_PSConstantBuffer.resize((int)PixelConstantBufferType::End, nullptr);
 	m_VSConstantBuffer.resize((int)VertexConstantBufferType::End, nullptr);
+#ifdef ROOTEX_EDITOR
```
Comment (IronicallySerious) [inline rootex/core/renderer/materials/textured_material.cpp:22]: Newline before

File: rootex/core/renderer/materials/textured_material.cpp [hunk 13]
```diff
@@ -1,18 +1,27 @@
...
 
+#include "resource_loader.h"
+
 #include "renderer/shader_library.h"
 #include "framework/systems/light_system.h"
 #include "framework/systems/render_system.h"
 #include "renderer/texture.h"
 
-TexturedMaterial::TexturedMaterial(Ref<Texture> diffuseTexture)
+TexturedMaterial::TexturedMaterial(String& imagePath)
```
Comment (IronicallySerious) [inline rootex/core/renderer/materials/textured_material.cpp:10]: Take in a `const String&`

File: rootex/core/renderer/materials/textured_material.cpp [hunk 14]
```diff
@@ -1,18 +1,27 @@
...
     : Material(ShaderLibrary::GetDiffuseShader())
-    , m_DiffuseTexture(diffuseTexture)
+    //, m_DiffuseTexture(diffuseTexture)
     , m_DiffuseShader(reinterpret_cast<DiffuseShader*>(m_Shader))
+    //, m_ImageFile(imageRes)
 {
+	m_TypeName = "Textured Material";
+	m_ImageFile = ResourceLoader::CreateImageResourceFile(imagePath);
```
Comment (IronicallySerious) [inline rootex/core/renderer/materials/textured_material.cpp:17]: Better done in initialization list

File: rootex/core/renderer/materials/textured_material.cpp [hunk 15]
```diff
@@ -30,6 +39,16 @@ void TexturedMaterial::setVSConstantBuffer(const VSDiffuseConstantBuffer& consta
 	Material::setVSConstantBuffer<VSDiffuseConstantBuffer>(constantBuffer, m_VSConstantBuffer[(int)VertexConstantBufferType::Model], 1u);
 }
 
+Material* TexturedMaterial::CreateDefault()
+{
+	return new TexturedMaterial(String("rootex/assets/rootex.png"));
```
Comment (IronicallySerious) [inline rootex/core/renderer/materials/textured_material.cpp:44]: No need for String()

File: rootex/core/renderer/materials/textured_material.cpp [hunk 16]
```diff
@@ -38,3 +57,57 @@ void TexturedMaterial::bind()
 	setPSConstantBuffer({ LightSystem::GetSingleton()->getLights() });
 	setPSConstantBuffer(PSDiffuseConstantBufferMaterial({ 0.6f, 30.0f, { 0.0f, 0.0f } }));
```
Comment (IronicallySerious) [inline rootex/core/renderer/materials/textured_material.cpp:58]: This

File: rootex/core/renderer/material.h [hunk 17]
```diff
@@ -21,6 +23,13 @@ class Material
 	Material();
 	virtual ~Material() = default;
 	virtual void bind();
+	// TODO- Make this purely virtual
+	virtual void draw() = 0;
+	String getFileName() { return m_FileName; };
+	String getTypeName() { return m_TypeName; };
+	String getFullName() { return m_FileName + " - " + m_TypeName; };
+	void setFileName(String& fileName) { m_FileName = fileName; };
+	virtual JSON::json getJSON() const = 0;
```
Comment (IronicallySerious) [inline rootex/core/renderer/material.h:32]: Define getJSON() here that serializes m_FileName and m_Type. Now in ColorMaterial::getJSON() and other children of Material, do
```cpp
JSON::json ColorMaterial getJSON()
{
    JSON::json& j = Material::getJSON();
    // Add some things to `j` that only belong to ColorMaterial and not the parent class.
    return j;
}
```

File: rootex/core/renderer/materials/color_material.h [hunk 18]
```diff
@@ -16,11 +16,22 @@ class ColorMaterial : public Material
...
+
 	void setPSConstantBuffer(const PSSolidConstantBuffer& constantBuffer);
 	void setVSConstantBuffer(const VSSolidConstantBuffer& constantBuffer);
 
+	static Material* CreateDefault();
+	static Material* Create(const JSON::json& materialData);
+
 	void bind() override;
+	JSON::json getJSON() const override;
+
```
Comment (IronicallySerious) [inline rootex/core/renderer/materials/color_material.h:32]: Extra newline

File: rootex/core/renderer/materials/color_material.h [hunk 19]
```diff
@@ -16,11 +16,22 @@ class ColorMaterial : public Material
...
+
+#ifdef ROOTEX_EDITOR
+	void draw() override;
```
Comment (IronicallySerious) [inline rootex/core/renderer/materials/color_material.h:35]: `virtual void draw() override`

File: rootex/core/renderer/materials/textured_material.cpp [hunk 20]
```diff
@@ -38,3 +57,57 @@ void TexturedMaterial::bind()
...
 }
+
+JSON::json TexturedMaterial::getJSON() const
+{
+	JSON::json j;
+
+	j["type"] = "Textured Material";
```
Comment (IronicallySerious) [inline rootex/core/renderer/materials/textured_material.cpp:65]: Check the comment on Material::getJSON()

File: rootex/core/renderer/materials/textured_material.cpp [hunk 21]
```diff
@@ -38,3 +57,57 @@ void TexturedMaterial::bind()
...
+			}
+			else
+			{
+				WARN("Cannot assign a non-image file to a texture");
+			}
+		}
+		ImGui::EndDragDropTarget();
+	}
+}
+#endif // ROOTEX_EDITOR
```
Comment (IronicallySerious) [inline rootex/core/renderer/materials/textured_material.cpp:113]: Newline in the end

File: rootex/core/renderer/materials/textured_material.cpp [hunk 22]
```diff
@@ -38,3 +57,57 @@ void TexturedMaterial::bind()
...
+	m_DiffuseTexture = texture;
+}
+
+#ifdef ROOTEX_EDITOR
+#include "imgui.h"
+#include "imgui_stdlib.h"
+void TexturedMaterial::draw()
+{
+	ImGui::Text("Textured Material");
+	m_ImagePathUI = m_ImageFile->getPath().string();
```
Comment (IronicallySerious) [inline rootex/core/renderer/materials/textured_material.cpp:84]: I think this line would never let you edit the path from the ImGui::InputText box. Check if this works

File: rootex/core/renderer/materials/textured_material.cpp [hunk 23]
```diff
@@ -38,3 +57,57 @@ void TexturedMaterial::bind()
...
+	}
+
+	if (ImGui::BeginDragDropTarget())
+	{
+		if (const ImGuiPayload* payload = ImGui::AcceptDragDropPayload("Resource Drop"))
+		{
+			const char* payloadFileName = (const char*)payload->Data;
+			FilePath payloadPath(payloadFileName);
+			FilePath ext = payloadPath.extension();
+			if (ext == ".jpg" || ext == ".png" || ext == ".jpeg")
```
Comment (IronicallySerious) [inline rootex/core/renderer/materials/textured_material.cpp:101]: Remind me when you see this comment. I have made a solution that will not hardcode the extension names everywhere. I will have to do changes in this area when I rebase

File: rootex/core/renderer/materials/textured_material.h [hunk 24]
```diff
@@ -22,12 +29,20 @@ class TexturedMaterial : public Material
 		Material,
 		End
 	};
-	TexturedMaterial(Ref<Texture> diffuseTexture);
+	TexturedMaterial(String& imagePath);
```
Comment (IronicallySerious) [inline rootex/core/renderer/materials/textured_material.h:32]: Take in a `const String&`

File: rootex/framework/components/visual/cpu_particles_visual_component.cpp [hunk 25]
```diff
@@ -10,48 +8,40 @@
...
+		    componentData["angularVelocity"]["z"],
+		    componentData["angularVelocity"]["w"] },
+		{ componentData["colorBegin"]["r"],
+		    componentData["colorBegin"]["g"],
+		    componentData["colorBegin"]["b"],
+		    componentData["colorBegin"]["a"] },
+		{ componentData["colorEnd"]["r"],
+		    componentData["colorEnd"]["g"],
+		    componentData["colorEnd"]["b"],
+		    componentData["colorEnd"]["a"] },
```
Comment (IronicallySerious) [inline rootex/framework/components/visual/cpu_particles_visual_component.cpp:25]: What happened here?

File: rootex/framework/components/visual/model_visual_component.cpp [hunk 26]
```diff
@@ -14,28 +14,20 @@ Component* ModelVisualComponent::Create(const JSON::json& componentData)
...
-
 	return modelVisualComponent;
 }
 
 Component* ModelVisualComponent::CreateDefault()
 {
 	ModelVisualComponent* modelVisualComponent = new ModelVisualComponent(
 	    RenderPassMain,
-	    Ref<Material>(new ColorMaterial()), //change
+	    MaterialLibrary::GetDefaultMaterial(), //change
```
Comment (IronicallySerious) [inline rootex/framework/components/visual/model_visual_component.cpp:28]: Remove comment

File: rootex/framework/components/visual/model_visual_component.h [hunk 27]
```diff
@@ -32,8 +31,7 @@ class ModelVisualComponent : public VisualComponent
 	virtual void postRender() override;
 
 	void setVisualModel(VisualModelResourceFile* newModel);
-	void setMaterial(Ref<Material> material);//change
-	void setColor(const Color& color) { m_Color = color; };
+	void setMaterial(Ref<Material>& material);//change
```
Comment (IronicallySerious) [inline rootex/framework/components/visual/model_visual_component.h:34]: Remove comment

File: game/assets/materials/green.rmat [hunk 28]
```diff
@@ -0,0 +1,9 @@
+{
+    "type":"Color Material",
```
Comment (IronicallySerious) [inline game/assets/materials/green.rmat:2]: Should we not write "ColorMaterial"?

File: rootex/core/renderer/material_library.cpp [hunk 29]
```diff
@@ -0,0 +1,110 @@
...
+void MaterialLibrary::LoadMaterials()
+{
+	for (auto& material : OS::GetFilesInDirectory("game/assets/materials/"))
+	{
+		if (material.extension() == ".rmat")
+		{
+			String materialName = material.filename().string();
+			TextResourceFile* materialFile = ResourceLoader::CreateNewTextResourceFile("game/assets/materials/" + materialName);
+			const JSON::json materialJSON = JSON::json::parse(materialFile->getString());
+			s_Materials[materialName] = { (String)materialJSON["type"], {} };
```
[the remark this answers is not in the document]
Comment (sin3point14) [reply rootex/core/renderer/material_library.cpp:21]: Create and CreateDefault are store in `m_MaterialDatabase`, in this step only the list of available materials is retrieved, an object is created only when it is requested for

File: rootex/core/renderer/material_library.cpp [hunk 30]
```diff
@@ -0,0 +1,110 @@
...
+Ref<Material> MaterialLibrary::GetDefaultMaterial()
+{
+	if (Ref<Material> lockedMaterial = s_Materials["DEFAULT"].second.lock())
+	{
+		return lockedMaterial;
+	}
+	else
+	{
+		Ref<Material> material(ColorMaterial::CreateDefault());
+		material->setFileName(String("DEFAULT"));
```
[the remark this answers is not in the document]
Comment (sin3point14) [reply rootex/core/renderer/material_library.cpp:65]: the parameter of the function is a `String&`

[the remark this answers is not in the document]
Comment (IronicallySerious) [reply rootex/core/renderer/material_library.cpp:65]: Change the parameter of `setFileName()` to `const String&` and then no need to do String("Default")

File: rootex/core/renderer/materials/textured_material.cpp (hunk 16 above)
Comment (sin3point14) [reply rootex/core/renderer/materials/textured_material.cpp:58]: Do you mean the hardcoded values?

File: rootex/core/renderer/materials/textured_material.cpp (hunk 22 above)
Comment (sin3point14) [reply rootex/core/renderer/materials/textured_material.cpp:84]: This works, `setTexture(image)` changes `m_ImageFile`

File: rootex/framework/components/visual/cpu_particles_visual_component.cpp (hunk 25 above)
Comment (sin3point14) [reply rootex/framework/components/visual/cpu_particles_visual_component.cpp:25]: VS doing, VS things

File: game/assets/materials/green.rmat (hunk 28 above)
Comment (sin3point14) [reply game/assets/materials/green.rmat:2]: oh no........

File: rootex/core/renderer/materials/color_material.cpp [hunk 31]
```diff
@@ -1,9 +1,11 @@
 #include "color_material.h"
 
-#include "renderer/shader_library.h"
 #include "framework/systems/render_system.h"
+#include "renderer/shader_library.h"
 
-ColorMaterial::ColorMaterial()
+ColorMaterial::ColorMaterial(const Color& color)
+	: m_Color(color)
+    , Material(ShaderLibrary::GetDefaultShader(), ColorMaterial::s_MaterialName)
```
Comment (IronicallySerious) [inline rootex/core/renderer/materials/color_material.cpp:8]: Tab

File: rootex/core/renderer/materials/color_material.cpp [hunk 32]
```diff
@@ -19,8 +21,40 @@ void ColorMaterial::setVSConstantBuffer(const VSSolidConstantBuffer& constantBuf
 	Material::setVSConstantBuffer<VSSolidConstantBuffer>(constantBuffer, m_VSConstantBuffer[(int)VertexConstantBufferType::Model], 1u);
 }
 
+Material* ColorMaterial::CreateDefault()
+{
+	return new ColorMaterial({0.5f, 0.5f, 0.5f, 1.0f});
```
Comment (IronicallySerious) [inline rootex/core/renderer/materials/color_material.cpp:26]: Tab

**Comment by IronicallySerious** [comment]
LGTM

**Pull request merged**